test/test_id_model.cpp

```diff
@@ -935,9 +1123,113 @@ TEST_F(IdModelTest, LoopPromotion6) {
+  ASSERT_EQ(id84->name(), 84);
+  auto id85 = getChildId(id81, 1);
+  ASSERT_EQ(id85->name(), 85);
+  auto id86 = getChildId(id81, 1, 1);
+  ASSERT_EQ(id86->name(), 86);
+  auto id87 = getChildId(id83, 1);
+  ASSERT_EQ(id87->name(), 87);
+  auto id88 = getChildId(id83, 1, 1);
+  ASSERT_EQ(id88->name(), 88);
+
+  // Check Step 3 results. See the design doc for the expected results
+  std::vector<std::pair<std::unordered_set<Val*>, IterDomain*>>
+      s3_reference_map = {
+          // 1, 2, 3, 46, 6, 7, 41, 8, 9, 10, 35, 36, 11, 12, 61, 15, 16, 56,
+          // 17, 18, 29, 30, 79, 80 -> 79
+          {std::unordered_set<Val*>{
+               tv1->getRootDomain().at(0),
+               tv2->getRootDomain().at(0),
+               tv2->getRootDomain().at(1),
+               getChildId(tv2->getRootDomain().at(0), 1),
+               tv4->getRootDomain().at(0),
+               tv4->getRootDomain().at(1),
+               getChildId(tv4->getRootDomain().at(0), 1),
+               tv5->getRootDomain().at(0),
+               tv5->getRootDomain().at(1),
+               tv5->getRootDomain().at(2),
+               getChildId(tv5->getRootDomain().at(0), 1),
+               getChildId(tv5->getRootDomain().at(2), 1),
+               tv6->getRootDomain().at(0),
+               tv6->getRootDomain().at(1),
+               getChildId(tv6->getRootDomain().at(0), 1),
+               tv8->getRootDomain().at(0),
+               tv8->getRootDomain().at(1),
+               getChildId(tv8->getRootDomain().at(0), 1),
+               tv9->getRootDomain().at(0),
+               tv9->getRootDomain().at(1),
+               tv9->getRootDomain().at(2),
+               getChildId(tv9->getRootDomain().at(0), 1),
+               getChildId(tv9->getRootDomain().at(0), 2),
+               id79,
+               id80},
+           id80},
+          // 71, 47, 42, 37, 62, 57, 31, 81, 83 -> 83
+          {std::unordered_set<Val*>{
+               getChildId(tv1->getRootDomain().at(0), 1),
+               getChildId(tv2->getRootDomain().at(0), 2),
+               getChildId(tv4->getRootDomain().at(0), 2),
+               getChildId(tv5->getRootDomain().at(0), 3),
+               getChildId(tv6->getRootDomain().at(0), 2),
+               getChildId(tv8->getRootDomain().at(0), 2),
+               getChildId(tv9->getRootDomain().at(0), 3),
+               id81,
+               id83},
+           id83},
+          // 73, 49, 44, 39, 64, 59, 33, 85, 87 -> 87
+          {std::unordered_set<Val*>{
+               tv1->axis(0),
+               tv2->axis(0),
+               tv4->axis(0),
+               tv5->axis(0),
+               tv6->axis(0),
+               tv8->axis(0),
+               tv9->axis(0),
+               id85,
+               id87},
+           id87},
+          // 48 -> 43
+          {std::unordered_set<Val*>{tv2->axis(2)}, tv4->axis(2)},
+          // 50 -> 45
+          {std::unordered_set<Val*>{tv2->axis(1)}, tv4->axis(1)},
+          // 40, 88 -> 88
```

Comment:
Why 88 is grouped with 40, instead of 34? Is this arbitrary?

This makes me feel generally worried about doing a `mapThroughExpr` on the loop graph. For example, if we have:

```
I0   I1        I5   I6
 \  /           \   /
  I2      ->     I7
 /  \           /  \
I3 | I4        I8  I9
```
Then `I2` and `I7` will be in the same loop group. If we do a `mapThroughExpr` on this loop group forward, will we map `I4` and `I9`, because they are equivalent exprs?

I think mathematically, `mapThroughExpr` is assuming the "congruence" property: if x == y, then f(x) == f(y), but unfortunately loop graph is not a graph with congruence property, so now I am deeply worried about this line: https://github.com/NVIDIA/Fuser/blob/4df003a526561987b1be085c5dd286ddd964d87d/csrc/id_model/id_model.cpp#L1112

BTW: I learnt the word "congruence" from https://doi.org/10.1145/3434304, thanks @jacobhinkle for sharing this paper with me.

Reply:
> Does this address your concern?

Yes, it does. Thank you!

So this is the line that adds the replayed ID to some loop group?
https://github.com/NVIDIA/Fuser/blob/3281aa10727af3e01c974f6f13288d0c157bb776/csrc/id_model/id_model.cpp#L996-L997

Reply:
Yes.